app/domain/authentication/authn_azure/validate_application_identity.rb

```diff
@@ -6,122 +6,155 @@ module AuthnAzure
+        raise Err::RoleMissingConstraint, resource_type unless resource
       end
 
-      # validates that the application identity doesn't include logical constraint
+      # validates that the application identity doesn't include logical resource constraint
       # combinations (e.g user_assigned_identity & system_assigned_identity)
       def validate_constraint_combinations
-        identifiers = %i(user_assigned_identity system_assigned_identity)
+        identifiers = %w(user-assigned-identity system-assigned-identity)
 
-        identifiers_constraints = application_identity.constraints.keys & identifiers
+        identifiers_constraints = application_identity.resource_types & identifiers
         unless identifiers_constraints.length <= 1
-          raise Errors::Authentication::IllegalConstraintCombinations, annotation_type_constraints(identifiers_constraints)
+          raise Errors::Authentication::IllegalConstraintCombinations, identifiers_constraints
         end
       end
 
-      def validate_token_identity_matches_annotations
-        application_identity.constraints.each do |constraint|
-          annotation_type  = constraint[0].to_s
-          annotation_value = constraint[1]
-          unless annotation_value == @token_identity[annotation_type.to_sym]
-            raise Err::InvalidApplicationIdentity, annotation_type_constraint(annotation_type)
-          end
+      # xms_mirid is a term in Azure to define a claim that describes the resource that holds the encoding of the instance's
+      # among other details the subscription_id, resource group, and provider identity needed for authorization.
+      # xms_mirid is one of the fields in the JWT token. This function will extract the relevant information from
+      # xms_mirid claim and populate a representative hash with the appropriate fields.
+      def extract_resources_from_token
+        @resources_from_token = [
+          AzureResource.new(
+            resource_type: "subscription-id",
+            resource_value: xms_mirid.subscriptions
+          ),
+          AzureResource.new(
+            resource_type: "resource-group",
+            resource_value: xms_mirid.resource_groups
+          )
+        ]
+
+        # determine which identity is provided in the token. If the token is
+        # issued to a user-assigned identity then we take the identity name.
+        # If the token is issued to a system-assigned identity then we take the
+        # Object ID of the token.
+        if xms_mirid.providers.include? "Microsoft.ManagedIdentity"
+          @resources_from_token.push(
+            AzureResource.new(
+              resource_type: "user-assigned-identity",
+              resource_value: xms_mirid.providers.last
+            )
+          )
+        else
+          @resources_from_token.push(
+            AzureResource.new(
+              resource_type: "system-assigned-identity",
+              resource_value: @oid_token_field
+            )
+          )
```

Comment:
I see another missing concept here, in both the above and in what you are pulling out of the annotations, and that is the idea of `ResourceSet`.  Afaict from the code, a valid ResourceSet consists of:
1. subscription-id
2. resource-group
3. one of either `user-assigned-identity` or `system-identity`.

You can then create / validate such objects from either a token or from annotations.

You could them replace code like this:

```ruby
       application_identity.resources.each do |resource_from_role|
          @resources_from_token.each do |resource_from_token|
            if resource_from_token.resource_type == resource_from_role.resource_type &&
              resource_from_token.resource_value != resource_from_role.resource_value
                raise Err::InvalidApplicationIdentity, resource_from_role.resource_type
            end
          end
        end
```
with something like:
```ruby
if token_resource_set != annotation_resource_set
  raise Err::InvalidApplicationIdentity
end
```
and define a `mismatches(other_resource_set)` method on it, if it's important to include where they don't match.

I think changes like these would help clarify the high level logic, which I think is still simpler than it seems in this version of the code.

Reply:
I am not sure i understand your suggestion here but i just want to note that it's not really a `!=` check as they don't need to match. The logic says that the `token_resource_set` should be a super-set of the `annotations_resource_set`.

Reply:
So you may not want to use `!=` as the operator, but my point should stand.  Instead of a custom `==` operator you can create a `subset?(other)` method.  The only question is if the idea of `ResourceSet` (or another more appropriate name if you have one) is really at play here.  It looks like it is.   But let me know what else was unclear.

Reply:
yes, it seems correct. just wanted to verify.

Reply:
do you mean something like this:
```
module Authentication
  module AuthnAzure

    class ResourceSet

      attr_reader :resources

      def self.from_role(role_annotations:, service_id:, azure_resource_types:, logger:)
        resources = azure_resource_types.each_with_object([]) do |resource_type, resources|
          resource_value = resource_value(role_annotations, service_id, resource_type)
          if resource_value
            resources.push(
              AzureResource.new(
                type: resource_type,
                value: resource_value
              )
            )
          end
        end

        new(resources: resources, logger: logger)
      end
      
      def self.from_token(xms_mirid_token_field:, oid_token_field:, logger:)
        xms_mirid = XmsMirid.new(xms_mirid_token_field)
        
        resources ||= [
          AzureResource.new(
            type: "subscription-id",
            value: xms_mirid.subscriptions
          ),
          AzureResource.new(
            type: "resource-group",
            value: xms_mirid.resource_groups
          )
        ]

        # determine which identity is provided in the token. If the token is
        # issued to a user-assigned identity then we take the identity name.
        # If the token is issued to a system-assigned identity then we take the
        # Object ID of the token.
        if xms_mirid.providers.include? "Microsoft.ManagedIdentity"
          resources.push(
            AzureResource.new(
              type: "user-assigned-identity",
              value: xms_mirid.providers.last
            )
          )
        else
          resources.push(
            AzureResource.new(
              type: "system-assigned-identity",
              value: oid_token_field
            )
          )
        end
        logger.debug(Log::ExtractedResourceRestrictionsFromToken.new)

        new(resources: resources, logger: logger)
      end
      
      def valid?
        validate_permitted_scope
        validate_required_constraints_exist
        validate_constraint_combinations
      end
      
      def subset?(other_resource_set)
        @resources.each do |resource|
          other_resource_set.each do |other_resource|
            if other_resource.type == resource.type &&
              other_resource.value != resource.value
              raise Err::InvalidResourceRestrictions, resource.type
            end
          end
        end
      end
      
      private

      def initialize(resources:, logger:)
        @resources = resources
        @logger           = logger
      end

      # check the `service-id` specific resource first to be more granular
      def resource_value role_annotations, service_id, resource_type
        annotation_value(role_annotations, "authn-azure/#{service_id}/#{resource_type}") ||
          annotation_value(role_annotations, "authn-azure/#{resource_type}")
      end

      def annotation_value role_annotations, name
        annotation = role_annotations.find { |a| a.values[:name] == name }

        # return the value of the annotation if it exists, nil otherwise
        if annotation
          @logger.debug(LogMessages::Authentication::RetrievedAnnotationValue.new(name))
          annotation[:value]
        end
      end

      # validating that the annotations listed for the Conjur resource align with the permitted Azure constraints
      def validate_permitted_scope
        validate_prefixed_permitted_annotations("authn-azure/")
        validate_prefixed_permitted_annotations("authn-azure/#{@service_id}/")
      end

      # check if annotations with the given prefix is part of the permitted list
      def validate_prefixed_permitted_annotations prefix
        @logger.debug(LogMessages::Authentication::ValidatingAnnotationsWithPrefix.new(prefix))

        prefixed_annotations(prefix).each do |annotation|
          annotation_name = annotation[:name]
          next if prefixed_permitted_constraints(prefix).include?(annotation_name)
          raise Errors::Authentication::ConstraintNotSupported.new(annotation_name.gsub(prefix, ""), AZURE_RESOURCE_TYPES)
        end
      end

      def prefixed_annotations prefix
        role_annotations.select do |a|
          annotation_name = a.values[:name]

          annotation_name.start_with?(prefix) &&
            # verify we take only annotations from the same level
            annotation_name.split('/').length == prefix.split('/').length + 1
        end
      end

      # add prefix to all permitted constraints
      def prefixed_permitted_constraints prefix
        AZURE_RESOURCE_TYPES.map { |k| "#{prefix}#{k}" }
      end

      def validate_required_constraints_exist
        validate_resource_constraint_exists "subscription-id"
        validate_resource_constraint_exists "resource-group"
      end

      def validate_resource_constraint_exists resource_type
        resource = resource_restrictions.resources.find { |a| a.type == resource_type }
        raise Err::RoleMissingConstraint, resource_type unless resource
      end

      # validates that the resource restrictions do include logical resource constraint
      # combinations (e.g user_assigned_identity & system_assigned_identity)
      def validate_constraint_combinations
        identifiers = %w(user-assigned-identity system-assigned-identity)

        identifiers_constraints = resource_restrictions.resources.map(&:type) & identifiers
        unless identifiers_constraints.length <= 1
          raise Errors::Authentication::IllegalConstraintCombinations, identifiers_constraints
        end
      end

    end
  end
end
```

if so, it doesn't do much more than move the logic from `validate_resource_restrictions` to this class.

Reply:
ok I looked into it and I don't think it will be an easy change. In Azure it works in a way (although it mostly moves logic from `validate_resource_restrictions.rb` to this new class) but in k8s it will be a harder refactor. It is not conceptually wrong to do it in k8s but the refactor will need to go deeper for that, because of the way the k8s authenticator validates the resource restrictions.

So at this point I'd rather leave it the way it is now.

Reply:
I would suggest that instead of trying to find the abstraction first, that you make the code for each authenticator as clear and simple as possible.

If, after that, there is still commonality, you can factor it out.  

If not, then it was probably a premature abstraction.  

In general, it's dangerous to let your abstractions dictate how you code, especially when they don't match your mental model.  Instead, they should arise naturally out of the code.  I still maintain that in the case of these authenticators the individual code for each authenticator is not yet simple enough, and so I would first work to make that happen.

Reply:
please see this commit with the changes: https://github.com/cyberark/conjur/pull/1580/commits/2beaffa81f83a5cfb9fce2da05ab07a18132e723

do you think it is easier to understand now?